records_mover/url/resolver.py

```diff
@@ -1,14 +1,109 @@
+from urllib.parse import urlparse
 from .base import BaseFileUrl, BaseDirectoryUrl
-from . import FileUrl, DirectoryUrl
-from typing import Any
+import inspect
+from typing import Callable, Optional, Dict, Any, Type, Union, TYPE_CHECKING
+if TYPE_CHECKING:
+    import google.cloud.storage  # noqa
+    import boto3.session  # noqa
+import logging
 
 
+logger = logging.getLogger(__name__)
+
+CreatesUrls = Callable[..., Union[BaseFileUrl, BaseDirectoryUrl]]
+
+
+directory_url_ctors: Dict[str, Union[Type[BaseDirectoryUrl], CreatesUrls]] = {}
+file_url_ctors: Dict[str, Union[Type[BaseFileUrl], CreatesUrls]] = {}
+
+
+def init_urls() -> None:
+    try:
+        from .s3.s3_url import S3Url
+    except ModuleNotFoundError:
+        logger.debug('No S3 support', exc_info=True)
+        S3Url = None  # type: ignore
+    try:
+        from .gcs.gcs_file_url import GCSFileUrl
+        from .gcs.gcs_directory_url import GCSDirectoryUrl
+    except ModuleNotFoundError:
+        logger.debug('No Google Cloud Storage support', exc_info=True)
+        GCSFileUrl = None  # type: ignore
+        GCSDirectoryUrl = None  # type: ignore
+    from .filesystem import FilesystemDirectoryUrl, FilesystemFileUrl
+    from .http import HttpFileUrl
+    if len(directory_url_ctors) == 0:
+        if S3Url is not None:
+            directory_url_ctors['s3'] = S3Url
+        if GCSDirectoryUrl is not None:
+            directory_url_ctors['gs'] = GCSDirectoryUrl
+        directory_url_ctors['file'] = FilesystemDirectoryUrl
+    if len(file_url_ctors) == 0:
+        if S3Url is not None:
+            file_url_ctors['s3'] = S3Url
+        if GCSFileUrl is not None:
+            file_url_ctors['gs'] = GCSFileUrl
+        file_url_ctors['file'] = FilesystemFileUrl
+
+        file_url_ctors['http'] = HttpFileUrl
+        file_url_ctors['https'] = HttpFileUrl
+
+
+# TODO: Instead of three functions, how about passing in an object that includes an interface?
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker